docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,435 @@
+    
+    impl AsRef<&str> for Denom {
+        fn as_ref(&self) -> &str{
+            self.0.as_str()
+        }
+    }
+
+    impl FromStr for Denom{
+        fn from_str(val: &str) -> Self {
+            Self(val.to_string())
+        }
+    }
+
+    /// Coin defines a token with a denomination and an amount.
+    #[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+    pub struct Coin {
+        /// Denomination
+        pub denom: DenomTrace,
+
+        /// Amount
+        pub amount: U256,
+    }
+
+    pub struct TracePrefix {
+        pub port_id: PortId,
+        pub channel_id: ChannelId
+    }
+
+    /// This struct and it's implementations should help identifying denomination traces
+    pub struct DenomTrace {
+        /// Full denomination path
+        pub trace_path: Vec<TracePrefix>,
+        /// Base denimination
+        pub base_denom: Denom
+    }
+
+    impl DenomTrace {
+        /// Returns the full denom path
+        fn get_full_denom_path(&self) -> String {
+            self.trace_path.into_iter().fold(String::new(), |acc, val| {
+                let next = Self::get_denom_prefix(&val.port_id, &val.channel_id);
+                format!("{}{}", acc, next)
+            })
+        }
+        /// IBCDenom a coin denomination for an ICS20 fungible token in the format
+        /// 'ibc/trace_path/base_denom'. If the trace is empty, it will return the base denomination.
+        fn ibc_denom(&self) -> String {
+            format!("ibc/{}", self.get_full_denom_path())
+        }
+        /// Returns the prefix for this trace
+        fn get_prefix(&self) -> String {
+            DenomTrace::get_denom_prefix(self.trace_path[0].port_id, self.trace_path[0].channel_id)
+        }
+
+        /// Returns the prefix for this trace
+        pub fn has_prefix(denom: String, prefix: String) -> bool {
+            denom.starts_with(prefix)
+        }
+
+        /// get_denom_prefix returns the receiving denomination prefix
+        pub fn get_denom_prefix(port_id: &PortId, channel_id: &ChannelId) -> String {
+            format!("{}/{}/", port_id, channel_id)
+        }
+
+
+        /// get_prefixed_denom returns the denomination with the port_id and channel_id prefixed
+        pub fn get_prefixed_denom(port_id: &PortId, channel_id: &ChannelId, base_denom: String) -> String {
+            format!("{}/{}/{}", port_id, channel_id, base_denom)
+        }
+        
+        pub fn parse_denom_trace(denom: &str) -> Result<Self, ICS20Error> {
+            
+        }
+    }
+
+    impl FromStr for DenomTrace {
+        type Error = ICS20Error
+        fn from_str(val: &str) -> Result<Self, Self::Error> {
+            DenomTrace::parse_denom_trace(val)
+        }
+    }
+
+    /// sender_chain_is_source returns false if the denomination originally came
+    /// from the receiving chain and true otherwise.
+    fn sender_chain_is_source(source_port: &PortId, source_channel: &ChannelId, denom: String) -> bool {
```

Comment:
We should combine `sender_chain_is_source()` and `receiver_chain_is_source()` into one `get_source_chain() -> SourceChain` (names TBD, please don't use those necessarily), where
```rust
enum SourceChain {
    Sender,
    Receiver,
}
```
This increases readability, as match statements will be used where both possibilities are made explicit, instead of being captured in an `if/else` statement.